### PR TITLE
cpp: add support for function static initializers

### PR DESCRIPTION
**Pull request by lowlander**
To support thread-safe, one-time initialization of function scope
variables that have non-trivial constructors, the compiler will
generate calls to the following functions, __cxa_guard_acquire,
__cxa_guard_release, and __cxa_guard_abort.

Those functions make sure that the initialization is only done
one time, by using a global zephyr mutex.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

### CODE REVIEW

**Review by pabigot — CHANGES_REQUESTED** [review]
From the description this is associated only with function static initializers, not global static initializers, right?

This one I'm not comfortable with yet.  It's coupled to the Zephyr initialization process: AIUI you can't use `k_mutex` until the boot process has reached at least [POST_KERNEL](https://docs.zephyrproject.org/latest/reference/drivers/index.html#initialization-levels).

For function static initializers I'd be concerned about what happens if a function that has a static initializer gets invoked before initialization had proceeded that far.  I suspect these functions need a check and `k_panic()` call if it's too soon.

For global static initializers I think we might need something that ensures that whatever invokes these functions is deferred to that state.

I'm leery of supporting static initializers only for one of the two contexts: it'd confuse people.

I also would like input from a kernel developer like @andrewboie, who I recall had (somewhere I can't find) commented about missing startup capabilities to support this sort of thing.  I remember looking at it once and thinking that the C++ ctor functions were being called later than strictly necessary.

All that aside we do at minimum need a test for this, either standalone or integrated with `tests/application_development/libcxx`, verifying that the initializer is called at the correct time (i.e. verify not-yet-called before `f()`, called when first in `f()`, not called again when back in `f()`).

**Comment by lowlander** [comment]
> From the description this is associated only with function static initializers, not global static initializers, right?

Yes this only applies to static variables inside a function, C++ guarantees that they are only initialized one time. For POD this doesn't need any locking, for example;
```
void f() {
  static int i = 42;
}
```
does not use the guard functions. But as soon as you have a object with a non-trivial constructor the guard functions are called. 
```
struct S {
  S(){}
  int i{42};
}

void f() {
  S s;
}
```
Here the guard functions are called.

> 
> This one I'm not comfortable with yet. It's coupled to the Zephyr initialization process: AIUI you can't use `k_mutex` until the boot process has reached at least [POST_KERNEL](https://docs.zephyrproject.org/latest/reference/drivers/index.html#initialization-levels).
> 
> For function static initializers I'd be concerned about what happens if a function that has a static initializer gets invoked before initialization had proceeded that far. I suspect these functions need a check and `k_panic()` call if it's too soon.

> 
> For global static initializers I think we might need something that ensures that whatever invokes these functions is deferred to that state.

> 
> I'm leery of supporting static initializers only for one of the two contexts: it'd confuse people.
> 
> I also would like input from a kernel developer like @andrewboie, who I recall had (somewhere I can't find) commented about missing startup capabilities to support this sort of thing. I remember looking at it once and thinking that the C++ ctor functions were being called later than strictly necessary.
> 
> All that aside we do at minimum need a test for this, either standalone or integrated with `tests/application_development/libcxx`, verifying that the initializer is called at the correct time (i.e. verify not-yet-called before `f()`, called when first in `f()`, not called again when back in `f()`).

Maybe a "stupid" question, but is it even possible to call a C++ function before the kernel init is done? Because Zephyr itself is not allowed (by convention) to use C++, so the first time a C++ function (with a static non-trivial object) can be called is as result of a global constructor, or in/after main. Or am I missing some strange usecase here. But even if that usecase exists it is already broken when it can call C++ before global constructors are run. So if global constructors are run after the kernel init is done, the static initializer are never going to be run before the kernel init is done.

A "simple" ASSERT with a test for the init level should be enough I think. 

I can add a test case, but aren't the tests always run after the kernel init is already done ?

**Comment by lowlander** [comment]
PS: CI complains about "Codeowners" what should I do there ?

**Comment by pabigot** [comment]
> Maybe a "stupid" question, but is it even possible to call a C++ function before the kernel init is done? Because Zephyr itself is not allowed (by convention) to use C++

Well, "convention".  Some people want to write drivers in C++, or somebody might just want to register something with `SYS_INIT` and not be aware of the limitation, so we should take reasonable steps to detect mis-use.  I guess `__ASSERT` would be OK, though generally `CONFIG_DEBUG=n` so the failure might be hidden.

> I can add a test case, but aren't the tests always run after the kernel init is already done ?

Yes, but you're specifically adding support for static local variables that aren't zero- or constant-initialized.  So the test is already running at the time the initialization occurs and you can verify that it's being called at that point.

> PS: CI complains about "Codeowners" what should I do there ?

I've added #19773 so the warning should go away when it's merged.

File: subsys/cpp/cpp_guard.c
Comment (vanwinkeljan) [inline subsys/cpp/cpp_guard.c]: Maybe move this check into the CMakeLists.txt as there is not need to compile this file if `CONFIG_LIB_CPLUSPLUS` is set.

Comment (lowlander) [reply subsys/cpp/cpp_guard.c]: That was just copied from the other cpp files in that directory, maybe make a separate cleanup commit?

**Comment by lowlander** [comment]
> > Maybe a "stupid" question, but is it even possible to call a C++ function before the kernel init is done? Because Zephyr itself is not allowed (by convention) to use C++
> 
> Well, "convention". Some people want to write drivers in C++, or somebody might just want to register something with `SYS_INIT` and not be aware of the limitation, so we should take reasonable steps to detect mis-use. I guess `__ASSERT` would be OK, though generally `CONFIG_DEBUG=n` so the failure might be hidden.

There is probably no way to detect it at compile time? But putting the check after the initial (lock free) check the impact of an "always on" check should be very minimal. After a quick browse in the docu I could not find a `get_init_level()` function, can someone tell me how to check the init level at runtime?

> 
> > I can add a test case, but aren't the tests always run after the kernel init is already done ?
> 
> Yes, but you're specifically adding support for static local variables that aren't zero- or constant-initialized. So the test is already running at the time the initialization occurs and you can verify that it's being called at that point.
> 

I was thinking about how to test the negative case, so when using it before kernel init is done, to check if the "panic" is triggered.

> > PS: CI complains about "Codeowners" what should I do there ?
> 
> I've added #19773 so the warning should go away when it's merged.

Ok, will rebase when that is in master.

**Comment by vanwinkeljan** [comment]
Adding a simple test would be a good idea but adding it  to `tests/application_development/libcxx` would have no effect as `CONFIG_LIB_CPLUSPLUS` is set. There is a build only none-libc++ test `tests/application_development/cxx` maybe this one could be extended to run the tests.

**Comment by lowlander** [comment]
@pabigot I am just wondering, when k_mutex isn't allowed to be used yet, the kernel is still in single thread mode, isn't it? If so, than the mutex locking can be skipped. And only when multi threading (and mutex use) is possible the mutex lock can be taken.

**Comment by lowlander** [comment]
For the qemu_x86_long the test with CONFIG_LIB_CPLUSPLUS fails;  undefined reference to `__cxa_guard_release'

Seems like the compiler and/or stdlib aren't correct for that platform ?

**Comment by vanwinkeljan** [comment]
> Seems like the compiler and/or stdlib aren't correct for that platform ?

There is a toolchain issue for qemu_x86_long, see https://github.com/zephyrproject-rtos/zephyr/pull/18464#discussion_r332656452

**Comment by lowlander** [comment]
Throwing exceptions seems completely broken on all except ARM platforms, they all seem to "exit" and than hang qemu even with a catch(...)

**Comment by lowlander** [comment]
@pabigot I can't find a way to check what init level the kernel is in, it seems that in kernel/init.c the current level isn't stored.  So i am not sure how to add an ASSERT or panic to check if the mutex may be used.  Will the k_mutex_lock fail in a defined way when it is called when not yet allowed? because than the existing ASSERT would catch it.

**Comment by pabigot** [comment]
> @pabigot I can't find a way to check what init level the kernel is in, it seems that in kernel/init.c the current level isn't stored.

No, it's all done with linker magic.  The order in  `kernel/init.c` at `bg_thread_main` is:
* `POST_KERNEL` initializations
* `APPLICATION` initializations
* C++ static constructors.

I think it makes more sense for the C++ ctor initialization to be moved before the boot banner and `APPLICATION` initializations, because conceptually that's more where it would be run normally, but again we need some insight from the folks who put it so late (specifically, why?).

Then you could either use `SYS_INIT()` to register something that's run at the highest priority in the `APPLICATION` level, or you could use a static C++ constructor to set up the state.

Or, perhaps the best approach, pull the C++ `__do_foo_aux()` calls out to a separate function, and add setup to support function static initialization after global static initialization completes.

**Comment by andrewboie** [comment]
> I think it makes more sense for the C++ ctor initialization to be moved before the boot banner and APPLICATION initializations, because conceptually that's more where it would be run normally, but again we need some insight from the folks who put it so late (specifically, why?).

More like, why do you want it this early?
To me this hook should be pulled in at the APPLICATION init level.
You just need this stuff to have run before main() starts, no?

**Comment by lowlander** [comment]
@nashif any idea why CI times out on this PR?

**Comment by lowlander** [comment]
OK now CI somehow doesn't time out anymore, and all checks pass, what else should be done to get this PR include?

@vanwinkeljan Is it OK to do the `#ifdef` to cmake part in a separate PR ? Because zephyr-cmake isn't really my thing, and I don't want to break anything, and the `#ifdef` setup is used by the other existing files also.

@pabigot and @andrewboie I agree with Andrew that drivers and other stuff that runs before global constructors are run should not use C++. And with that assumption this PR should be fine, or do you see any other problems?

**Comment by pabigot** [comment]
Somehow this seems to have grown beyond its original scope of adding function static initializers.

I guess the complete reorganization and extension of the tests area is ok in principle, but I think it will cause conflicts with several previous C++-related PRs that aren't getting any attention, like #18464, #19735, #19507, and perhaps others.  Several of those are from @vanwinkeljan who should chime in, but I think those should go in first, since among other things they demonstrate how to properly decouple Zephyr's C++ support from newlib's.

**Comment by lowlander** [comment]
@vanwinkeljan can I do anything (testing or so) to help move forward those PR's @pabigot mentioned?

**Comment by vanwinkeljan** [comment]
@lowlander a review of #18464 would be welcome

**Comment by vanwinkeljan** [comment]
@lowlander #18464 was merged

**Comment by lowlander** [comment]
Waiting for #20406 to be fixed so I can rebase and retest the x86_64 tests.

**Comment by progwolff** [comment]
What's the status of this? Can I help in some way?

**Comment by lowlander** [comment]
@progwolff not sure what the zephyr plans are when it comes to C++, for the moment this PR seems on hold, I tried to keep it conflict free for a while but stopped doing that.  The patch itself should be easy to apply to your own code if you need it (that's what I did).

File: tests/application_development/cpp/cpp_compile/testcase.yaml
Comment (henrikbrixandersen) [inline tests/application_development/cpp/cpp_compile/testcase.yaml:8]:
```suggestion
    platform_allow: native_posix
```

**Comment by lowlander** [comment]
This PR might become obsolete by the work done for issue #18554

**Comment by alexanderwachter** [comment]
@lowlander @pabigot What is the state of this PR? We need that for our project. Seems that it is in a mergeable state.

**Comment by lowlander** [comment]
> @lowlander @pabigot What is the state of this PR? We need that for our project. Seems that it is in a mergeable state.

I have no idea what the state of C++ for Zephyr is and of the usefulness of this PR, it could also be handled in the toolchain/libs.

**Comment by alexanderwachter** [comment]
@tejlmand what do you think? For C++ support, we definitely need those guard functions.

**Review by tejlmand — CHANGES_REQUESTED** [review]
> @tejlmand what do you think? For C++ support, we definitely need those guard functions.

The general look on this seems good, and it's almost ready to go in, imho.
Only have one minor comment regarding CMake, as we are now using `find_package(...)` for locating Zephyr.

File: tests/application_development/cpp/static_vars/CMakeLists.txt
Comment (lowlander) [reply tests/application_development/cpp/static_vars/CMakeLists.txt:4]: The PR is already a bit older (also still has u32/u8) , I will update it to master later this week.

**Comment by lowlander** [comment]
@tejlmand & @alexanderwachter I removed a few tests that were not related to the static initializers and rebased on master, should be ok now I think.

**Review by alexanderwachter — APPROVED** [review]

File: subsys/cpp/cpp_guard.c
Comment (andrewboie) [inline subsys/cpp/cpp_guard.c:10]: User mode code won't be able to use this `k_mutex`. `sys_mutex` would allow this however.

Is there a specific need for priority inheritance? If not I think a semaphore would be better, it will pull in less code on small systems. A `sys_sem` can also be used from user mode if it is placed in the default memory domain.

I suggest:
```
#include <app_memory/app_memdomain.h>

static K_APP_DMEM(z_libc_partition) SYS_SEM_DEFINE(heap_sem, 1, 1);
```

And use `sys_sem_take()`/`sys_sem_give()` to implement the critical section.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]
I'd like this to be workable for user mode threads, see suggested change inline.

File: subsys/cpp/cpp_guard.c
Comment (lowlander) [reply subsys/cpp/cpp_guard.c:10]: @andrewboie 

> Is there a specific need for priority inheritance?

Hmmm it is a "global" lock so it think priority inversion could happen. If thread A starts the initialization of a static var and gets preempted before it is ready it will block thread B that might want to initialize another static var. And if than a thread with a prio in between A and B starts an endless loop, the high prio thread B never gets done. So unless I am thinking in the wrong direction, a mutex would be better ?

File: subsys/cpp/cpp_guard.c
Comment (pabigot) [inline subsys/cpp/cpp_guard.c:30]: Where is this API documented?  Is it specific to GCC?

Why is the guard object a `uint64_t`?  The GCC source uses `__guard` which it appears to typedef as `int`.

This needs to be documented somewhere so we have a clue for when other toolchains need to be supported.

File: subsys/cpp/cpp_guard.c
Comment (pabigot) [reply subsys/cpp/cpp_guard.c:10]: I agree priority inversion could be a problem, but whatever's done this should probably use the same priority behavior as devices.  See #26073.

**Review by pabigot — CHANGES_REQUESTED** [review]
I want to understand where this lock API came from and why its argument is typed the way it is: it's not obvious that it's correct. 
 This information should be documented in the module.

Otherwise I have no specific objections to this, we're a long way from C++ thread support in Zephyr and maybe this will get us closer.

File: subsys/cpp/cpp_guard.c
Comment (lowlander) [reply subsys/cpp/cpp_guard.c:30]: Clang seems to use uint64_t. After some more searching (see link) the ARM C++ ABI seems to say 32bit and the Itanium (and everybody else) seems to say 64bit. So this patch might need a `#ifdef` for arm (not sure what is with arm64) and the rest. 

But like I mentioned earlier, this is something that probably is better implemented in the toolchain (gcc, clang, what ever), but I have no real knowledge on what the state is, and if C++ is even a thing that the toolchain maintainers are interested in. 

https://android.googlesource.com/platform/bionic/+/c9a659c/libc/bionic/__cxa_guard.cpp

**Review by tejlmand — APPROVED** [review]

**Comment by alexanderwachter** [comment]
Seems that we have problems in native_posix:
```
valgrind ./build/zephyr/zephyr.elf
==121918== Memcheck, a memory error detector
==121918== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==121918== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==121918== Command: ./build/zephyr/zephyr.elf
==121918== 
==121918== Invalid read of size 1
==121918==    at 0x10F0CC: z_impl_k_mutex_lock (mutex.c:133)
==121918==    by 0x10CD5F: k_mutex_lock (kernel.h:674)
==121918==    by 0x10CD5F: __cxa_guard_acquire (cpp_guard.c:38)
==121918==    by 0x492B075: __future_category_instance (future.cc:65)
==121918==    by 0x492B075: std::future_category() (future.cc:79)
==121918==    by 0x48FB508: __static_initialization_and_destruction_0 (compatibility-thread-c++0x.cc:50)
==121918==    by 0x48FB508: _GLOBAL__sub_I_compatibility_thread_c__0x.cc (compatibility-thread-c++0x.cc:124)
==121918==    by 0x40112DA: call_init.part.0 (in /usr/lib32/ld-2.32.so)
==121918==    by 0x40113D9: _dl_init (in /usr/lib32/ld-2.32.so)
==121918==    by 0x40010F9: ??? (in /usr/lib32/ld-2.32.so)
==121918==  Address 0xe is not stack'd, malloc'd or (recently) free'd
```

**Comment by lowlander** [comment]
> Seems that we have problems in native_posix:

@alexanderwachter might be because of the way native posix uses the native libs and how it is linked ? 

I think the whole C++ thing can only work when the toolchains are helping, but C++ seems to have zero priority.

**Comment by alexanderwachter** [comment]
The problem is that we are using `k_mutex_lock()` before the kernel has initialized.
in mutex.c:
```C
mutex->owner_orig_prio = (mutex->lock_count == 0U) ? _current->base.prio : mutex->owner_orig_prio;
```
_current is zero in this case -> segfault.

```C
int __cxa_guard_acquire(uint64_t *guard_object)
{
	int res;

	if (init_has_run(guard_object)) {
		return 0;
	}
	if (_current) {
		res = k_mutex_lock(&guard_mutex, K_FOREVER);
	__ASSERT_NO_MSG(res == 0);
	}
	

	if (init_has_run(guard_object)) {
		if (_current) {
			k_mutex_unlock(&guard_mutex);
		}

		return 0;
	}

	return 1;
}
```
Works but is ugly

**Comment by lowlander** [comment]
> The problem is that we are using `k_mutex_lock()` before the kernel has initialized.
> in mutex.c:

Is this a native posix problem because it runs things before zephyr is started ?  When is `_dl_init ` called ?

**Comment by alexanderwachter** [comment]
Not sure. I think it is a general problem. Maybe @andrewboie knows more.

**Comment by r2r0** [comment]
What about using ```k_is_pre_kernel```?
If it returns false then mutex is not needed and if it returns true then mutex is required (and also is available and can be used).
This will make such guard independent on Zephyr initialization order.

**Comment by lowlander** [comment]
> What about using `k_is_pre_kernel`?
> If it returns false then mutex is not needed and if it returns true then mutex is required (and also is available and can be used).
> This will make such guard independent on Zephyr initialization order.

I don't think this has to do with Zephyr itself but with how native posix works, and C++ code is initialized before any zephyr code is run.

**Comment by andrewboie** [comment]
> Not sure. I think it is a general problem. Maybe @andrewboie knows more.

why do we need this before the POST_KERNEL phase?
It doesn't make any sense to me that POSIX is initializing itself before we're in main thread context, this might just a very simple fix to initialize it in POST_KERNEL

**Comment by lowlander** [comment]
> why do we need this before the POST_KERNEL phase?
> It doesn't make any sense to me that POSIX is initializing itself before we're in main thread context, this might just a very simple fix to initialize it in POST_KERNEL

I think the problem isn't zephyr's init order, but the way native posix is linked and when C++ code is called (at least it looks like that from the stack trace above) 

A real "fix" is probably only possible with toolchain support for C++.

**Pull request closed without merging**
